### ISSUE TITLE
[color-slider] Incorrect color

### ISSUE THREAD

**Issue by LeaVerou** [color-js/elements#22]
<img width="1024" alt="image" src="https://github.com/color-js/elements/assets/175836/81f44ad5-f73c-4bec-a600-fc8e44498751">

**Comment by DmitrySharabin** [comment]
Related to #32 and is fixed when that issue is fixed.

**Comment by LeaVerou** [comment]
Duplicate of #32

**Comment by DmitrySharabin** [comment]
Debugging this issue, I noticed that when we apply changes to the color stops, we use the default color space https://github.com/color-js/elements/blob/61be842e960d269a8eaef9e34797b689b626fd44/src/color-slider/color-slider.js#L208-L210 instead of the specified one.

The sequence is the following: (default) space (`sRGB`) → color stops (in `sRGB`) → space from the attribute (`HSL`). After that, color stops are not updating (!) and remain in the `sRGB` color space.

As Lea pointed out in other examples we discussed, the slider has the wrong gradient.

**Comment by DmitrySharabin** [comment]
Or they shouldn't be in the same space? 🤔

**Comment by DmitrySharabin** [comment]
In our discussion, Lea mentioned that the issue is not with the thumb color but the gradient. She said that CSS does not support a raw hue interpolation, only [`longer`, `shorter`, `increasing`, and `decreasing`](https://drafts.csswg.org/css-color-4/#hue-interpolation). So we need to *fake it* with `tesselateStops()`. Precisely, we need to detect if a) the color space is polar and b) if the hues of adjacent color stops are significantly far apart (>= 180).

So, I came up with the solution we might want to use (by adding it in [this block](https://github.com/color-js/elements/blob/61be842e960d269a8eaef9e34797b689b626fd44/src/color-slider/color-slider.js#L70-L82)):

```js
let farApart = false;
if (this.space.isPolar) {
	for (let i = 1; i < stops.length; i++) {
		let firstHue = stops[i - 1].get("h");
		let secondHue = stops[i].get("h");

		if (Math.abs(firstHue - secondHue) >= 180) {
			farApart = true;
			break;
		}
	}
}

if (!supported || farApart) {
	stops = this.tessellateStops({ steps: 3 });
}
```

Unfortunately, this didn't fix the issue, so I need to investigate it further.

**Comment by LeaVerou** [comment]
> As Lea pointed out in other examples we discussed, the slider has the wrong gradient.

In the example you showed me it had the wrong gradient. Here the gradient is definitely correct. Possibly different issues?

**Comment by DmitrySharabin** [comment]
> In the example you showed me it had the wrong gradient. Here the gradient is definitely correct. Possibly different issues?

Don't know. 🤷‍♂️ Will check. If so, I will file a separate issue.

**Comment by DmitrySharabin** [comment]
I worked at it a bit more.

[If we take a look at the screenshot](https://github.com/color-js/elements/issues/22#issue-2284350669) in the first comment, we'll find out that the current color is `hsl(162, 100%, 50%)`. I checked this color in some other apps to see what it looks like, and it turned out to look _exactly_ the same as in the screenshot.

If the gradient is correct, that means something is wrong with the `scales` we build based on color stops,

https://github.com/color-js/elements/blob/9506ca436a21bb8bf1dbc62c15ecb21b25bdd0af/src/color-slider/color-slider.js#L232-L245

since we get (the thumb) color from a scale we get based on the progress

https://github.com/color-js/elements/blob/9506ca436a21bb8bf1dbc62c15ecb21b25bdd0af/src/color-slider/color-slider.js#L165

**Comment by DmitrySharabin** [comment]
Yes, the issue is with `scales`—they are evaluated only once (and cached) when the `stops` attribute is handled (on the element initialization). They are evaluated based on colors in the `sRGB` color space. However, the `color` is evaluated in `HSL`, and the `scales` are interpreted as if they were evaluated in `HSL`, as well. But it's not the case.

If we set the `<color-slider>` `space` to `sRGB`, we'll get the correct result:

```html
<color-slider space="srgb" stops="red, yellow, lime, aqua, blue, magenta, red"></color-slider>
```

<img width="1288" alt="image" src="https://github.com/color-js/elements/assets/9166277/7a802f77-9a87-48c4-a1cd-53914a79e962">

**Comment by DmitrySharabin** [comment]
Found it!!!!

Here https://github.com/color-js/elements/blob/9506ca436a21bb8bf1dbc62c15ecb21b25bdd0af/src/color-slider/color-slider.js#L240 should be `outputSpace: this.space` instead of `space: this.space`!

**Comment by LeaVerou** [comment]
- `outputSpace` defaults to `space`: https://github.com/color-js/color.js/blob/main/src/interpolation.js#L151
- We don’t want to default to the color space of the colors specified, since we almost never want to interpolate in sRGB.

**Comment by DmitrySharabin** [comment]
Thanks to [this brilliant interactive page](https://colorjs.io/docs/interpolation) 🤩, I could find out why we face this issue—it's because of the interpolation method. `heu: "raw"` produces an incorrect result in our case. 

### Wrong

`Input`
```js
let color = new Color("magenta");
let color2 = color.range("red", {
	space: "hsl",
	hue: "raw"
});
color2(.7);
``` 
`Output`
<img width="198" alt="image" src="https://github.com/color-js/elements/assets/9166277/d4505de2-0378-4e26-b228-666e89166953">

### Correct

`Input`
```js
let color = new Color("magenta");
let color2 = color.range("red", {
	space: "hsl"
});
color2(.7);
```

`Output`
<img width="190" alt="image" src="https://github.com/color-js/elements/assets/9166277/039390b0-ff89-4a4a-8251-1ab48248eda2">

@LeaVerou, do you have any suggestions on what to do in this case?

**Comment by LeaVerou** [comment]
Once you tesselate stops that differ by less than 180 (or less than 90?), you won't need `raw` any more, so you can just remove it. :)
